pkg/operator/ceph/controller/version.go

```diff
@@ -47,3 +58,79 @@ func GetImageVersion(cephCluster cephv1.CephCluster) (*cephver.CephVersion, erro
 
 	return nil, errors.New("attempt to determine ceph version for the current cluster image timed out")
 }
+
+// DetectCephVersion loads the ceph version from the image and checks that it meets the version requirements to
+// run in the cluster
+func DetectCephVersion(rookImage, namespace, jobName string, ownerInfo *k8sutil.OwnerInfo, clientset kubernetes.Interface, cephClusterSpec *cephv1.ClusterSpec) (*version.CephVersion, error) {
+	cephImage := cephClusterSpec.CephVersion.Image
+	logger.Infof("detecting the ceph image version for image %s...", cephImage)
+	versionReporter, err := cmdreporter.New(
+		clientset,
+		ownerInfo,
+		jobName,
+		jobName,
+		namespace,
+		[]string{"ceph"},
+		[]string{"--version"},
+		rookImage,
+		cephImage,
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to set up ceph version job")
+	}
+
+	job := versionReporter.Job()
+	job.Spec.Template.Spec.ServiceAccountName = "rook-ceph-cmd-reporter"
+
+	// Apply the same placement for the ceph version detection as the mon daemons except for PodAntiAffinity
+	cephv1.GetMonPlacement(cephClusterSpec.Placement).ApplyToPodSpec(&job.Spec.Template.Spec)
+	job.Spec.Template.Spec.Affinity.PodAntiAffinity = nil
+
+	stdout, stderr, retcode, err := versionReporter.Run(detectCephVersionTimeout)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to complete ceph version job")
+	}
+	if retcode != 0 {
+		return nil, errors.Errorf(`ceph version job returned failure with retcode %d. `+
+			`stdout: %s. `+
+			`stderr: %s`,
+			retcode,
+			stdout,
+			stderr,
+		)
+	}
+
+	version, err := cephver.ExtractCephVersion(stdout)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to extract ceph version")
+	}
+	logger.Infof("detected ceph image version: %q", version)
+	return version, nil
+}
+
+func CurrentAndDesiredCephVersion(rookImage, namespace, jobName string, ownerInfo *k8sutil.OwnerInfo, context *clusterd.Context, cephClusterSpec *cephv1.ClusterSpec, clusterInfo *cephclient.ClusterInfo) (*version.CephVersion, *version.CephVersion, error) {
+	// Detect desired CephCluster version
+	desiredCephVersion, err := DetectCephVersion(rookImage, namespace, fmt.Sprintf("%s-detect-version", jobName), ownerInfo, context.Clientset, cephClusterSpec)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to detect ceph image version")
+	}
+
+	// Check the ceph version of the running monitors
+	runningMonDaemonVersion, err := client.LeastUptodateDaemonVersion(context, clusterInfo, config.MonType)
```

Comment:
Are there any potential problems with only checking the mon version here? Like if the file controller sees that mons are updated and assumes file doesn't need updated when it actually does need updated?

Reply:
The only purpose of this function is to wait for the mons to be the same version as the version in the image. If not, the reconciles requeue until this is true. When the version match, the reconcile runs normally and will update the spec with the new image, upgrading the daemons.